packages/framework/synthesize/src/dependencyContainer.ts

```diff
@@ -95,69 +103,94 @@ export class DependencyContainer implements IFluidDependencySynthesizer {
     }
 
     private generateOptional<T extends IFluidObject>(
+        base: AsyncOptionalFluidObjectProvider<FluidObjectKey<T>>,
         types: FluidObjectSymbolProvider<T>,
     ) {
-        const values: (keyof IFluidObject)[] = Object.values(types);
-        return Object.assign({}, ...Array.from(values, (t) => {
-            const provider = this.getProvider(t);
-            if (!provider) {
-                return { get [t]() { return Promise.resolve(undefined); } };
+        for(const key of Object.keys(types) as unknown as (keyof IFluidObject)[]) {
+            const provider = this.resolveProvider(key);
+            if(provider !== undefined) {
+                Object.defineProperty(
+                    base,
+                    key,
+                    provider,
+                );
             }
-
-            return this.resolveProvider(provider, t);
-        }));
+        }
     }
 
-    private resolveProvider<T extends keyof IFluidObject>(provider: FluidObjectProvider<T>, t: keyof IFluidObject) {
+    private resolveProvider<T extends keyof IFluidObject>(t: T): PropertyDescriptor | undefined {
+        // If we have the provider return it
+        const provider = this.providers.get(t);
+        if (provider === undefined) {
+            for(const parent of this.parents) {
+                // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
+                const sp = { [t]: t } as FluidObjectSymbolProvider<Pick<IFluidObject, T>>;
+                // eslint-disable-next-line @typescript-eslint/ban-types
+                const syn =  parent.synthesize<Pick<IFluidObject, T>,{}>(
+                    sp,
+                    {});
+                const descriptor = Object.getOwnPropertyDescriptor(syn, t);
+                if(descriptor !== undefined) {
+                    return descriptor;
+                }
+            }
+            return undefined;
+        }
+
         // The double nested gets are required for lazy loading the provider resolution
         if (typeof provider === "function") {
-            // eslint-disable-next-line @typescript-eslint/no-this-alias
-            const self = this;
             return {
-                get [t]() {
+                get() {
                     if (provider && typeof provider === "function") {
-                        return Promise.resolve(provider(self)).then((p) => {
+                        return Promise.resolve(this[IFluidDependencySynthesizer])
+                            .then(async (fds): Promise<any>=>provider(fds))
+                            .then((p) => {
+                                if (p) {
+                                    return p[t];
+                                }
+                            });
+                    }
+                },
+            };
+        }
+        return {
+                get() {
+                    if (provider) {
+                        return Promise.resolve(provider).then((p) => {
```

Comment:
I don't really understand what this is solving, but I see it's pre-existing

Reply:
provider is typed like `Promise<T> | T` wrapping in a resolve makes it always `Promise<T>`